framework/src/main/java/org/checkerframework/framework/util/FlowExpressionParseUtil.java

```diff
@@ -119,286 +85,180 @@ private static Pattern anchored(String regex) {
+            throw constructParserException(expression, "is an invalid expression");
+        }
+        Receiver result = getReceiverFromExpression(expr, context, localScope);
         if (result instanceof ClassName && !expression.endsWith("class")) {
             throw constructParserException(
                     expression, "a class name cannot terminate a flow expression string");
         }
         return result;
     }
 
-    private static FlowExpressions.Receiver parseHelper(
-            String expression, FlowExpressionContext context, TreePath path)
+    /**
+     * Converts the passed Javaparser Expression to a {@link Receiver}.
+     *
+     * @return the Receiver of the passed expression
+     */
+    private static Receiver getReceiverFromExpression(
+            Expression expr, FlowExpressionContext context, TreePath path)
             throws FlowExpressionParseException {
-        expression = expression.trim();
 
         ProcessingEnvironment env = context.checkerContext.getProcessingEnvironment();
         Types types = env.getTypeUtils();
 
-        if (isNullLiteral(expression, context)) {
-            return parseNullLiteral(types);
-        } else if (isIntLiteral(expression, context)) {
-            return parseIntLiteral(expression, types);
-        } else if (isLongLiteral(expression, context)) {
-            return parseLongLiteral(expression, types);
-        } else if (isFloatLiteral(expression, context)) {
-            throw constructParserException(
-                    expression,
-                    String.format("floating-point values '%s' cannot be parsed", expression));
-        } else if (isStringLiteral(expression, context)) {
-            return parseStringLiteral(expression, types, env.getElementUtils());
-        } else if (isThisLiteral(expression, context)) {
-            return parseThis(context);
-        } else if (isSuperLiteral(expression, context)) {
-            return parseSuper(expression, types, context);
-        } else if (isIdentifier(expression)) {
-            return parseIdentifier(expression, env, path, context);
-        } else if (isParameter(expression, context)) {
-            return parseParameter(expression, context);
-        } else if (isArray(expression)) {
-            return parseArray(expression, context, path);
-        } else if (isMethodCall(expression)) {
-            return parseMethodCall(expression, context, path, env);
-        } else if (isMemberSelect(expression)) {
-            return parseMemberSelect(expression, env, context, path);
-        } else if (isParentheses(expression, context)) {
-            return parseParentheses(expression, context, path);
-        } else {
-            String message;
-            if (expression.equals("#0")) {
-                message =
-                        "one should use \"this\" for the receiver or \"#1\" for the first formal parameter";
+        if (expr != null) {
+            if (expr.isNullLiteralExpr()) {
```

Comment:
This giant if/elseif/else statement can be re-written using a com.github.javaparser.ast.visitor.GenericVisitorWithDefaults.  The class would look something like this:

```
    public static class ExpressionToReceiverVisitor extends GenericVisitorWithDefaults<Receiver, FlowExpressionContext> {
        private final TreePath path;
        public ExpressionToReceiverVisitor(TreePath path) {
            this.path = path;
        }
        @Override
        public Receiver defaultAction(com.github.javaparser.ast.Node n, FlowExpressionContext arg) {
            // throw a parse exception saying something about an unsupported expressions type.
            return null;
        }

        @Override
        public Receiver visit(NullLiteralExpr n, FlowExpressionContext arg) {
              return new ValueLiteral(types.getNullType(), (Object) null);
        }
        // Add a visit method for each kind of Expression that we handle.
    }
```

Then this if/elseif/else can be replaced with   `expr.accept(new ExpressionToReceiverVisitor(path), context);`.
In the Visitor, don't call the private static methods, but rather copy the implementation of those methods to the proper visit method.

Reply:
(You'll need to pull and rebuild the stubparser, because I just pushed a fix for GenericVisitorWithDefaults.)

Reply:
If I replace evert `getXReceiver` method with  `visit` method inside `ExpressionToReceiverVisitor`, then some of them need to throw a `FlowExpressionParseException`. In this case they can't override the methods in `GenericVisitorWithDefaults`. One solution for this would be making `FlowExpressionParseException` an unchecked exception (extend `RuntimeException` instead of `Exception`). Should I do this?

Reply:
Also, can you please explain what you meant here:

> You'll need to pull and rebuild the stubparser, because I just pushed a fix for GenericVisitorWithDefaults.

I'm not sure what I need to do.

Reply:
I mean do a `git pull` in your clone of stubparser and rebuild.  (I think running   `./gradlew cloneAndBuildDependencies` in your Checker Framework directory will work, too.) 

>Update: Does it have something to do with the fact I need to provide an implementation for visit(StubUnit, A)?

Yes, that's what I fixed:
https://github.com/typetools/stubparser/commit/1737a1910b1d26544d8bbb527686d566361f4e3d

Reply:
> One solution for this would be making `FlowExpressionParseException` an unchecked exception (extend RuntimeException instead of Exception).

This would work, but we want it to be a checked exception so that clients catch it.  Another option would be to wrap the `FlowExpressionParseException` in a runtime exception.  For example,
```
throw new ParserRuntimeException( constructParserException(...));
```
The the parse method would catch the ParserRuntimeException and then throw the FlowExpressionParseException.

Or the  visitor could return a pair of a Receiver and an FlowExpressionParseException.  
```
class ExpressionToReceiverVisitor extends <Pair<Receiver, ExpressionToReceiverVisitor>, FlowExpressionContext>
```
Then either the receiver is null or the exception is null.  If the visitor returns a non-null exception, then the parse method will throw it.  

I haven't thought about which approach would be better.